packages/workflow/src/worker-interface.ts

```diff
@@ -0,0 +1,248 @@
+ */
+import ivm from 'isolated-vm';
+import { coresdk } from '@temporalio/proto';
+import { ActivityOptions, ApplyMode, ExternalDependencyFunction, WorkflowInfo } from './interfaces';
+import { consumeCompletion, ExternalCall, state } from './internals';
+import { tsToMs, msToTs } from './time';
+import { alea } from './alea';
+import { IsolateExtension, HookManager } from './promise-hooks';
+import { DeterminismViolationError, IllegalStateError } from './errors';
+
+export { ExternalCall };
+
+export function setRequireFunc(fn: Exclude<typeof state['require'], undefined>): void {
+  state.require = fn;
+}
+
+export function overrideGlobals(): void {
+  const global = globalThis as any;
+  // Mock any weak reference holding structures because GC is non-deterministic.
+  // WeakRef is implemented in V8 8.4 which is embedded in node >=14.6.0.
+  // Workflow developer will get a meaningful exception if they try to use these.
+  global.WeakMap = function () {
+    throw new DeterminismViolationError('WeakMap cannot be used in workflows because v8 GC is non-deterministic');
+  };
+  global.WeakSet = function () {
+    throw new DeterminismViolationError('WeakSet cannot be used in workflows because v8 GC is non-deterministic');
+  };
+  global.WeakRef = function () {
+    throw new DeterminismViolationError('WeakRef cannot be used in workflows because v8 GC is non-deterministic');
+  };
+
+  const OriginalDate = globalThis.Date;
+
+  global.Date = function () {
+    return new OriginalDate(state.now);
+  };
+
+  global.Date.now = function () {
+    return state.now;
+  };
+
+  global.Date.prototype = OriginalDate.prototype;
+
+  global.setTimeout = function (cb: (...args: any[]) => any, ms: number, ...args: any[]): number {
+    const seq = state.nextSeq++;
+    state.completions.set(seq, {
+      resolve: () => cb(...args),
+      reject: () => undefined /* ignore cancellation */,
+    });
+    state.commands.push({
+      startTimer: {
+        timerId: `${seq}`,
+        startToFireTimeout: msToTs(ms),
+      },
+    });
+    return seq;
+  };
+
+  global.clearTimeout = function (handle: number): void {
+    state.nextSeq++;
+    state.completions.delete(handle);
+    state.commands.push({
+      cancelTimer: {
+        timerId: `${handle}`,
+      },
+    });
+  };
+
+  // state.random is mutable, don't hardcode its reference
+  Math.random = () => state.random();
+}
+
+/** Mock DOM element for Webpack dynamic imports */
+export interface MockElement {
+  getAttribute(name: string): unknown;
+  setAttribute(name: string, value: unknown): void;
+}
+
+/** Mock document object for Webpack dynamic imports */
+export interface MockDocument {
+  head: {
+    // Ignored
+    appendChild(): void;
+  };
+  getElementsByTagName(): MockElement[];
+  createElement(): MockElement;
+}
+
+/**
+ * Create a mock document object with mimimal required attributes to support Webpack dynamic imports
+ */
+export function mockBrowserDocumentForWebpack(): MockDocument {
```

Comment:
Why is it necessary to fake a browser environment? It may cause some issues with outside modules. Most should be fine since you can't do I/O from workflows anyway, but could be a bit brittle.

Reply:
We fake it just for webpack to store user code in the mock element.
If you see where this is used in the isolate builder, you'll notice that we delete the document object immediately after webpack is done using it.